Add guidance on API documentation (#4160)
kubernetes-sigs/gateway-api: site-src/contributing/devguide.md

To add a new tool, use `go get -tool -modfile tools/go.mod the.tool.repo/toolname@version`
and tidy the specific module with `go mod tidy -modfile=tools/go.mod`.

To execute the new tool, use `go tool -modfile=tools/go.mod toolname`.

## API Documentation

When writing API documentation (the `godoc` for API fields and structures) it should be done

>  in a meaningful and concise way, where information are provided for the different 

Reviewer: ```suggestion
 in a meaningful and concise way, where information is provided for the different 
```

>  Gateway API personas (Ian, Chihiro and Ana) without leaking implementation details.

Reviewer 2: nit: potentially link to https://gateway-api.sigs.k8s.io/concepts/roles-and-personas/#key-roles-and-personas for the personas?

Author: good call, added

>  Gateway API personas (Ian, Chihiro and Ana) without leaking implementation details.

Reviewer 3: Is there a reason these lines are indented?

Author: nope, probably something from my vscode

The implementation details are still important for a Gateway API implementation
developer, and they should still be provided but without being exposed on the
CRD generation, that can end leaking to users on a diverse set of ways, like

> on Gateway API documentation website, or via `kubectl explain`.

Reviewer: ```suggestion
The implementation details are still important for Gateway API implementation
developers, and should still be provided, but we need to ensure they are not 
exposed in the generated CRDs. That can end up leaking to users in multiple ways,
such as on the Gateway API documentation website, or via `kubectl explain`.
```

Additionally, it is worth noticing that API documentation reflects on the CRD generation
size, which impacts directly on resource consumption like a maximum Kubernetes resource size 
(which is limited by etcd maximum value size) and avoiding problems with `last-applied-configuration` 

> annotation, when doing a client-side apply.

Reviewer 3: Is "maximum Kubernetes resource size" only one example of multiple impacts?  If there are other impacts, what are they?  If the Kubernetes resource size is the real issue here, can you delete "resource consumption like" to make the language more precise?
```suggestion
Additionally, it is worth noting that API documentation affects the size of generated CRD manifests,
which directly impacts the Kubernetes resource size, 
which is in turn limited by etcd maximum value size.  The situation is exacerbated by the `last-applied-configuration` 
annotation from client-side apply, which roughly doubles the size of the Kubernetes resource.
```

Author: from what I can think now, yes, this is the main (if not only) example of impact on Kubernetes

Author: (btw fixed)

> There are two kind of API documentations: 

Reviewer 3: ```suggestion
There are two kind of API documentation:
```

Author: fixed on my local branch

> * User facing - MUST define how a user should be consuming an API and its field, on a concise way.

Reviewer 3: ```suggestion
* User facing - MUST define how a user should be consuming an API and its fields, in a concise way.
```

Author: fixed on my local branch

* Developer facing - MUST define how a controller should implement an API and its fields. 

### User facing Documentation

The API documentation, when meaningful, helps users of it on doing proper configuration

> in a way that Gateway API controllers react and configure the proxies the right way.

Reviewer: ```suggestion
The API documentation, when meaningful, helps users of it understand and create configuration
in a way that ensures that their Gateway API implementation does what they want it to do.
```

A good API documentation should cover:

> * What is the main feature of the API and Field - Eg.: "`Foo` allows configuring how a

Reviewer 4: nit/maybe not: is 

```
`Foo` allows ...
```

a useful user facing comment given a user writes `foo: ...` not `Foo: ...` in yaml?

Author: tbh I think I added this based on the kube-api-linter (long time ago...) but I never liked the enforcement of field naming like this.

That said, I think it is better to have the right field name and not the structure name as you mentioned, so changing here for "foo"

a header should be forwarded to backends"
* What is the support level of the field - Eg.: "Support: Core/Extended/Implementation Specific"
* Caveats of that field - Eg.: "Setting `Foo` field can have conflicts with `Bar` field, and in this
case it will be shown as a Condition". (we don't need to cover all the conditions).

In a simple way, a user reading the field documentation should understand, on one or two 
phrases what happens when the field is configured, what can be configured and what are 

> the impacts of that field

Reviewer 3: Is this paragraph basically a recap of the preceding list?  
```suggestion
In one or two phrases, API documentation should help a user who is reading the documentation understand
what happens when the field is configured, what values can be specified, and what any
additional important implications of setting the field may be.  
```

Author: yes, but reinforcing it.

Changed the text here as suggested

When adding a documentation, it is very important to remove your "Developer hat" 
and put yourself on a user that is trying to solve a problem: Does setting a field

> solves my needs? How can I use it?

Reviewer 3: ```suggestion
When adding a documentation, it is very important to remove your "Developer hat" 
and put yourself in the role of a user who is trying to solve a problem: Does setting a field
solve my needs? How can I use it?
```

Author: done

On an implementation, a user facing documentation belongs to the field documentation. Taking

> `Listeners`, one of the most complex fields as an example:

Reviewer 3: What do you mean by "on an implementation"?  This guidance is for the Gateway API spec, not for implementation-specific documentation, right?
```suggestion
The godoc for an API field contains user-facing documentation. Taking
`Listeners`, one of the most complex fields, as an example:
```

Author: yeah, my phrasing was confusing, thanks! added yours

```golang
// Listeners define logical endpoints that are bound on this Gateway's addresses.
// At least one Listener MUST be specified. When setting a Listener, conflicts can
// happen depending on its configuration like protocol, hostname and port, and in 
// this case a status condition will be added representing what was the conflict.
// 
// The definition of a Listener protocol implies what kind of Route can be attached 
// to it
Listeners []Listener `json:"listeners"`
```

We don't specify what are the Protocol types (saving this to the `Protocol` field),
what a hostname means, when a TLS configuration is required. All of these information
belongs to each field, so when a user does something like `kubectl explain gateway.spec.listeners`

> they will also get the information of each field.

Reviewer 3: ```suggestion
We don't specify here what the Protocol types are (this information goes in the `Protocol` field itself),
how a hostname is defined, or when TLS configuration is required. All of this information
belongs to the respective API fields, which are sub-fields of `Listener`.  Note that `kubectl explain` prints information about both
the specified field and its sub-fields, so when a user executes the `kubectl explain gateway.spec.listeners` command,
they get all of this information.
```

Author: done

### Developer facing documentation

Developer facing documentation helps during implementations to define the expected

> behavior of it, and should answer questions like:

Reviewer 3: ```suggestion
Developer-facing documentation helps implementers to define the expected
behavior and should answer questions such as the following:
```
Edit: Ah, you write "implementors" elsewhere.  I believe "implementers" is more common, but both are valid spellings, so better to pick your favorite spelling and use it consistently.

Author: yeah, went with implementers now

> * How that field should be reconciled?

Reviewer 3: ```suggestion
* How should that field be reconciled?
```

Author: fixed

* What conditions should be set during the reconciliation? 
* What should be validated during the reconciliation of that field?

In this case, as the API documentation serves as a guide for implementors on how 
their implementations should behave, it is very important to be as much verbose as
required to avoid any ambiguity. These information are used also to define expected
conformance behavior, and can even point to existing GEPs so a developer looking 
at it can know where to look for more references on what and why are those the expected

> behavior of this field.

Reviewer 3: ```suggestion
their implementations should behave, it is very important to be as verbose as
required to avoid any ambiguity. This information is used also to define expected
conformance behavior, and it can point to existing GEPs so a developer looking 
at it can know where to look for more references on what the expected
behavior of this field is and why.
```

Author: fixed

Still taking the `Listeners` field as an example, it does good definitions of situations 

> like:

Reviewer 3: ```suggestion
Continuing with the `Listeners` field as an example, it shows how API documentation can provide guidance for situations 
such as the following:
```

Author: fixed

* Two listeners have different protocols but the same hostname. Should this be a conflict?
* A listener of type `XXX` sets the field `TLS`. Is this a problem? How to expose this to 
the user?

> Because these information don't matter for a user, they should be hidden from the CRD/OpenAPI

Reviewer 3: ```suggestion
Because these details don't matter for a user, they should be hidden from the CRD/OpenAPI
```
Hm, but do these details really not matter for the user?  Seems like a user would care about when they can specify `TLS` or specify the same host name on distinct listeners.

Reviewer 3: On further thought, these are details that you might want to omit from the CRD for the sake of brevity (and to keep the CRD manifest within size restrictions) but that you would actually want in the website API reference, which isn't subject to etcd's limits and can be more comprehensive.  But that thought implies that we really have these three categories of godoc:

- Brief user-facing (in the CRD).  
- Detailed user-facing (on the website).  
- Developer-facing (in the Go definitions).  

Maybe this warrants a separate discussion from this PR.

Author: yes, agreed. Maybe we need these as part of a "complete API reference" vs the "user API reference" or something like that. 

For me right now what matters, from a UX is that you do a kubectl explain, you get a brief explanation of the field and valid values without getting into unnecessary details
